internal/internal_task_pollers.go

```diff
@@ -896,6 +916,18 @@ func newGetHistoryPageFunc(
 			}
 			return h, nil, nil
 		}
+
+		// atDecisionTaskCompletedEventID (startedEventID) is 0 for query task
+		// maxEventID is 0 for old version of the server
+		// TODO: Apply the check to decision tasks (remove the last condition)
+		// after validating maxEventID always equal to atDecisionTaskCompletedEventID (startedEventID).
+		// For now only apply to query task to be safe.
+		if size > 0 && maxEventID != 0 && h.Events[size-1].GetEventId() > maxEventID && atDecisionTaskCompletedEventID == 0 {
```

Comment:
nit: maybe have some functions and save some comments:
if isQueryTaskWithNextEventID(maxEventID, atDecisionTaskCompletedEventID) && shouldTruncateHistory(size, maxEventID) {
...
}